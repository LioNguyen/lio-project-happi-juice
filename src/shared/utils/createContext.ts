import type React from "react";
import { createContext as originCreateContext, useContext } from "react";

function createContext<T>(): [React.Provider<T>, () => T, React.Context<T>] {
  const context = originCreateContext<T>({} as T);

  return [context.Provider, () => useContext(context), context];
}

export { createContext };
