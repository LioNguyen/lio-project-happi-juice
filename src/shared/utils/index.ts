export * from './lazyLoad'
export * from './tailwind'
