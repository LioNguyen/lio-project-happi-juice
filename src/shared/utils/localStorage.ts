/**
 * Save data to localStorage
 * @param key - The key under which the data will be stored
 * @param value - The data to store, can be any serializable value
 */
function setLocalStorage<T>(key: string, value: T): void {
  try {
    const serializedValue =
      typeof value === 'string' ? value : JSON.stringify(value)

    localStorage.setItem(key, serializedValue)
  } catch (error) {
    console.error(`Error saving ${key} to localStorage:`, error)
  }
}

/**
 * Retrieve data from localStorage
 * @param key - The key of the stored data
 * @returns The parsed data or null if not found or error occurs
 */
function getLocalStorage<T>(key: string): T | null {
  try {
    const serializedValue = localStorage.getItem(key)

    if (!serializedValue || serializedValue === 'undefined') {
      return null
    }

    // Try to parse JSON
    try {
      return JSON.parse(serializedValue) as T
    } catch {
      // Return as string if JSON parsing fails
      return serializedValue as T
    }
  } catch (error) {
    console.error(`Error reading ${key} from localStorage:`, error)
    return null
  }
}

/**
 * Remove data from localStorage
 * @param key - The key to remove from localStorage
 */
function removeLocalStorage(key: string): void {
  try {
    localStorage.removeItem(key)
  } catch (error) {
    console.error(`Error removing ${key} from localStorage:`, error)
  }
}

/**
 * Clear all data from localStorage
 */
function clearLocalStorage(): void {
  try {
    localStorage.clear()
  } catch (error) {
    console.error('Error clearing localStorage:', error)
  }
}

export {
  clearLocalStorage,
  getLocalStorage,
  removeLocalStorage,
  setLocalStorage,
}
