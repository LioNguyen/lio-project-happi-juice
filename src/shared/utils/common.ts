import { format, isValid, Locale, parseISO } from 'date-fns'
import { enUS, vi } from 'date-fns/locale'
import _ from 'lodash'

import { IOrderItem } from '@/domains/order'
import { ISelectOption } from '@/shared/types'

interface INumberFormatOptions {
  decimals?: number
  thousandSeparator?: string
  decimalSeparator?: string
}

// ==================== Date Utilities ====================
/**
 * Parse various date formats into Date object
 * @category Date
 * @param {Date | string | undefined} value - Input date
 * @returns {Date | undefined} Parsed Date object or undefined if invalid
 * @example
 * parseDate("2024-02-26") // Date object
 * parseDate("invalid") // undefined
 */
function parseDate(value: Date | string | undefined): Date | undefined {
  if (!value) return undefined
  if (value instanceof Date) return isValid(value) ? value : undefined

  try {
    // Try ISO string parsing first
    const date = parseISO(value)
    if (isValid(date)) return date

    // Fallback to direct Date creation
    const normalDate = new Date(value)
    return isValid(normalDate) ? normalDate : undefined
  } catch {
    return undefined
  }
}

/**
 * Format a date string according to specified format
 * @category Date
 * @param {string} date - Date string to format
 * @param {string} [dateFormat] - Output format (defaults to "yyyy-MM-dd")
 * @returns {string} Formatted date string
 * @example
 * formatDateTime("2024-02-26") // "2024-02-26"
 * formatDateTime("2024-02-26", "dd/MM/yyyy") // "26/02/2024"
 */
function formatDateTime(date: string, dateFormat?: string) {
  return format(date, dateFormat || 'yyyy-MM-dd')
}

const localeMap: Record<string, Locale> = {
  vi: vi,
  en: enUS,
}

// ==================== Number & Price Utilities ====================
/**
 * Format a number with customizable separators and decimals
 * @category Number
 * @param {number} value - Number to format
 * @param {INumberFormatOptions} [options] - Formatting options
 * @returns {string} Formatted number string
 */
const formatNumber = (
  value: number,
  options?: INumberFormatOptions,
): string => {
  const {
    decimals = 0,
    thousandSeparator = ',',
    decimalSeparator = '.',
  } = options || {}

  const parts = value.toFixed(decimals).split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, thousandSeparator)

  return parts.join(decimalSeparator)
}

/**
 * Formats the price with thousand separators using locale-specific formatting
 * @category Price
 * @param {string} price - Price string to format
 * @returns {string} Formatted price string
 */
const formatPrice = (price: string) => {
  return Number(price).toLocaleString()
}

// ==================== Array Utilities ====================
/**
 * Transforms an array of select options into a sorted, joined string
 * @category Array
 * @param {ISelectOption[]} array - Array of select options
 * @param {string} [key] - Key to extract from each option
 * @param {string} [separator] - Separator for joining values
 * @returns {string} Sorted and joined string
 */
const transformToSortedString = (
  array: ISelectOption[],
  key: string = 'value',
  separator: string = ',',
): string => {
  return _.chain(array).map(key).sortBy().join(separator).value()
}

// ==================== Order Utilities ====================
/**
 * Calculates the total price for an array of order items
 * @category Order
 * @param {IOrderItem[]} items - Array of order items
 * @returns {number} Total price
 */
const calculateTotalPrice = (items: IOrderItem[]) => {
  return items.reduce((sum, item) => sum + item.price * item.quantity, 0)
}

export {
  // Date utils
  parseDate,
  formatDateTime,
  localeMap,

  // Number & Price utils
  formatNumber,
  formatPrice,

  // Array utils
  transformToSortedString,
  // Order utils
  calculateTotalPrice,
}
