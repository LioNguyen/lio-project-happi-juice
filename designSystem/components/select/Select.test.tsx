import { screen, within } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";

import { renderWithTheme } from "@/shared/utils";
import {
  Select,
  SelectContent,
  SelectGroup,
  SelectItem,
  SelectLabel,
  SelectSeparator,
  SelectTrigger,
  SelectValue,
} from ".";

// Mock lucide-react icons
vi.mock("lucide-react", () => ({
  Check: () => <div data-testid="check-icon" />,
  ChevronDown: () => <div data-testid="chevron-down-icon" />,
  ChevronUp: () => <div data-testid="chevron-up-icon" />,
}));

// Mock radix-ui select components
vi.mock("@radix-ui/react-select", () => ({
  Root: vi.fn(({ children }) => <div data-testid="select-root">{children}</div>),
  Trigger: vi.fn(({ className, children, ...props }) => (
    <button data-testid="select-trigger" className={className} {...props}>
      {children}
    </button>
  )),
  Value: vi.fn(({ children }) => <span data-testid="select-value">{children}</span>),
  Portal: vi.fn(({ children }) => <div data-testid="select-portal">{children}</div>),
  Content: vi.fn(({ className, children, ...props }) => (
    <div data-testid="select-content" className={className} {...props}>
      {children}
    </div>
  )),
  Viewport: vi.fn(({ className, children }) => (
    <div data-testid="select-viewport" className={className}>
      {children}
    </div>
  )),
  Item: vi.fn(({ className, children, ...props }) => (
    <div data-testid="select-item" className={className} {...props}>
      {children}
    </div>
  )),
  ItemText: vi.fn(({ children }) => <span data-testid="select-item-text">{children}</span>),
  ItemIndicator: vi.fn(({ children }) => <span data-testid="select-item-indicator">{children}</span>),
  Group: vi.fn(({ children }) => <div data-testid="select-group">{children}</div>),
  Label: vi.fn(({ className, children, ...props }) => (
    <div data-testid="select-label" className={className} {...props}>
      {children}
    </div>
  )),
  Separator: vi.fn(({ className, ...props }) => (
    <div data-testid="select-separator" className={className} {...props} />
  )),
  ScrollUpButton: vi.fn(({ className, ...props }) => (
    <div data-testid="select-scroll-up-button" className={className} {...props} />
  )),
  ScrollDownButton: vi.fn(({ className, ...props }) => (
    <div data-testid="select-scroll-down-button" className={className} {...props} />
  )),
  Icon: vi.fn(({ children }) => <span data-testid="select-icon">{children}</span>),
}));

describe("Select Component", () => {
  const renderSelect = (props = {}) => {
    return renderWithTheme(
      <Select>
        <SelectTrigger>
          <SelectValue placeholder="Select a fruit" />
        </SelectTrigger>
        <SelectContent {...props}>
          <SelectGroup>
            <SelectLabel>Fruits</SelectLabel>
            <SelectItem value="apple">Apple</SelectItem>
            <SelectItem value="banana">Banana</SelectItem>
            <SelectSeparator />
            <SelectItem value="orange">Orange</SelectItem>
          </SelectGroup>
        </SelectContent>
      </Select>,
    );
  };

  // Basic Rendering Tests
  it("should render all select components", () => {
    renderSelect();

    expect(screen.getByTestId("select-root")).toBeInTheDocument();
    expect(screen.getByTestId("select-trigger")).toBeInTheDocument();
    expect(screen.getByTestId("select-value")).toBeInTheDocument();
    expect(screen.getByTestId("select-content")).toBeInTheDocument();
    expect(screen.getByTestId("select-group")).toBeInTheDocument();
    expect(screen.getByTestId("select-label")).toBeInTheDocument();
    expect(screen.getAllByTestId("select-item")).toHaveLength(3);
    expect(screen.getByTestId("select-separator")).toBeInTheDocument();
  });

  // Snapshot Tests
  describe("Select Snapshots", () => {
    it("should match snapshot with default props", () => {
      const { container } = renderSelect();
      expect(container).toMatchSnapshot();
    });

    it("should match snapshot with custom content", () => {
      const { container } = renderWithTheme(
        <Select>
          <SelectTrigger>
            <SelectValue placeholder="Custom Select" />
          </SelectTrigger>
          <SelectContent>
            <SelectItem value="custom">Custom Item</SelectItem>
          </SelectContent>
        </Select>,
      );
      expect(container).toMatchSnapshot();
    });
  });

  // SelectTrigger Tests
  describe("SelectTrigger", () => {
    it("should render with correct classes", () => {
      renderSelect();

      const trigger = screen.getByTestId("select-trigger");
      expect(trigger).toHaveClass(
        "flex",
        "h-9",
        "w-full",
        "items-center",
        "justify-between",
        "whitespace-nowrap",
        "rounded-md",
        "border",
        "border-input",
        "bg-transparent",
        "px-3",
        "py-2",
        "text-sm",
        "shadow-sm",
        "ring-offset-background",
        "[&>span]:line-clamp-1",
      );
    });

    it("should render chevron-down icon", () => {
      renderSelect();
      expect(screen.getAllByTestId("chevron-down-icon")[0]).toBeInTheDocument();
    });
  });

  // SelectContent Tests
  describe("SelectContent", () => {
    it("should render with correct classes", () => {
      renderSelect();

      const content = screen.getByTestId("select-content");
      expect(content).toHaveClass(
        "relative",
        "z-50",
        "max-h-96",
        "min-w-[8rem]",
        "overflow-hidden",
        "rounded-md",
        "border",
        "bg-popover",
        "text-popover-foreground",
        "shadow-md",
      );
    });

    it("should render scroll buttons", () => {
      renderSelect();
      expect(screen.getByTestId("select-scroll-up-button")).toBeInTheDocument();
      expect(screen.getByTestId("select-scroll-down-button")).toBeInTheDocument();
    });
  });

  // SelectItem Tests
  describe("SelectItem", () => {
    it("should render with correct classes", () => {
      renderSelect();

      const items = screen.getAllByTestId("select-item");
      items.forEach(item => {
        expect(item).toHaveClass(
          "relative",
          "flex",
          "w-full",
          "cursor-default",
          "select-none",
          "items-center",
          "rounded-sm",
          "py-1.5",
          "pl-2",
          "pr-8",
          "text-sm",
          "outline-none",
          "focus:bg-secondary",
        );
      });
    });

    it("should render check icon when selected", () => {
      renderWithTheme(
        <Select defaultValue="apple">
          <SelectContent>
            <SelectItem value="apple">Apple</SelectItem>
          </SelectContent>
        </Select>,
      );

      expect(screen.getByTestId("check-icon")).toBeInTheDocument();
    });
  });

  // SelectLabel Tests
  describe("SelectLabel", () => {
    it("should render with correct classes", () => {
      renderSelect();

      const label = screen.getByTestId("select-label");
      expect(label).toHaveClass("px-2", "py-1.5", "text-sm", "font-semibold");
    });
  });

  // SelectSeparator Tests
  describe("SelectSeparator", () => {
    it("should render with correct classes", () => {
      renderSelect();

      const separator = screen.getByTestId("select-separator");
      expect(separator).toHaveClass("-mx-1", "my-1", "h-px", "bg-muted");
    });
  });

  // ScrollButton Tests
  describe("ScrollButton Components", () => {
    it("should render scroll up button with icon", () => {
      renderSelect();
      const scrollUpButton = screen.getByTestId("select-scroll-up-button");
      expect(scrollUpButton).toBeInTheDocument();
      const upIcon = within(scrollUpButton).getByTestId("chevron-up-icon");
      expect(upIcon).toBeInTheDocument();
    });

    it("should render scroll down button with icon", () => {
      renderSelect();
      const scrollDownButton = screen.getByTestId("select-scroll-down-button");
      expect(scrollDownButton).toBeInTheDocument();
      const downIcon = within(scrollDownButton).getByTestId("chevron-down-icon");
      expect(downIcon).toBeInTheDocument();
    });

    it("should have correct styling classes", () => {
      renderSelect();
      const scrollButtons = [
        screen.getByTestId("select-scroll-up-button"),
        screen.getByTestId("select-scroll-down-button"),
      ];

      scrollButtons.forEach(button => {
        expect(button).toHaveClass("flex", "cursor-default", "items-center", "justify-center", "py-1");
      });
    });
  });
});
