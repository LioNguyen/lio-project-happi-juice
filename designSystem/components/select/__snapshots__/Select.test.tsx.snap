// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select Component > Select Snapshots > should match snapshot with custom content 1`] = `
<div>
  <div
    data-testid="select-root"
  >
    <button
      class="flex h-9 w-full items-center justify-between whitespace-nowrap rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1"
      data-testid="select-trigger"
    >
      <span
        data-testid="select-value"
      />
      <span
        data-testid="select-icon"
      >
        <div
          data-testid="chevron-down-icon"
        />
      </span>
    </button>
    <div
      data-testid="select-portal"
    >
      <div
        class="relative z-50 max-h-96 min-w-[8rem] overflow-hidden rounded-md border bg-popover text-popover-foreground shadow-md data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 data-[side=bottom]:translate-y-1 data-[side=left]:-translate-x-1 data-[side=right]:translate-x-1 data-[side=top]:-translate-y-1"
        data-testid="select-content"
        position="popper"
      >
        <div
          class="flex cursor-default items-center justify-center py-1"
          data-testid="select-scroll-up-button"
        >
          <div
            data-testid="chevron-up-icon"
          />
        </div>
        <div
          class="p-1 h-[var(--radix-select-trigger-height)] w-full min-w-[var(--radix-select-trigger-width)]"
          data-testid="select-viewport"
        >
          <div
            class="relative flex w-full cursor-default select-none items-center rounded-sm py-1.5 pl-2 pr-8 text-sm outline-none focus:bg-secondary focus:text-accent-foreground data-[disabled]:pointer-events-none data-[disabled]:opacity-50"
            data-testid="select-item"
            value="custom"
          >
            <span
              class="absolute right-2 flex h-3.5 w-3.5 items-center justify-center"
            >
              <span
                data-testid="select-item-indicator"
              >
                <div
                  data-testid="check-icon"
                />
              </span>
            </span>
            <span
              data-testid="select-item-text"
            >
              Custom Item
            </span>
          </div>
        </div>
        <div
          class="flex cursor-default items-center justify-center py-1"
          data-testid="select-scroll-down-button"
        >
          <div
            data-testid="chevron-down-icon"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Select Component > Select Snapshots > should match snapshot with default props 1`] = `
<div>
  <div
    data-testid="select-root"
  >
    <button
      class="flex h-9 w-full items-center justify-between whitespace-nowrap rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1"
      data-testid="select-trigger"
    >
      <span
        data-testid="select-value"
      />
      <span
        data-testid="select-icon"
      >
        <div
          data-testid="chevron-down-icon"
        />
      </span>
    </button>
    <div
      data-testid="select-portal"
    >
      <div
        class="relative z-50 max-h-96 min-w-[8rem] overflow-hidden rounded-md border bg-popover text-popover-foreground shadow-md data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 data-[side=bottom]:translate-y-1 data-[side=left]:-translate-x-1 data-[side=right]:translate-x-1 data-[side=top]:-translate-y-1"
        data-testid="select-content"
        position="popper"
      >
        <div
          class="flex cursor-default items-center justify-center py-1"
          data-testid="select-scroll-up-button"
        >
          <div
            data-testid="chevron-up-icon"
          />
        </div>
        <div
          class="p-1 h-[var(--radix-select-trigger-height)] w-full min-w-[var(--radix-select-trigger-width)]"
          data-testid="select-viewport"
        >
          <div
            data-testid="select-group"
          >
            <div
              class="px-2 py-1.5 text-sm font-semibold"
              data-testid="select-label"
            >
              Fruits
            </div>
            <div
              class="relative flex w-full cursor-default select-none items-center rounded-sm py-1.5 pl-2 pr-8 text-sm outline-none focus:bg-secondary focus:text-accent-foreground data-[disabled]:pointer-events-none data-[disabled]:opacity-50"
              data-testid="select-item"
              value="apple"
            >
              <span
                class="absolute right-2 flex h-3.5 w-3.5 items-center justify-center"
              >
                <span
                  data-testid="select-item-indicator"
                >
                  <div
                    data-testid="check-icon"
                  />
                </span>
              </span>
              <span
                data-testid="select-item-text"
              >
                Apple
              </span>
            </div>
            <div
              class="relative flex w-full cursor-default select-none items-center rounded-sm py-1.5 pl-2 pr-8 text-sm outline-none focus:bg-secondary focus:text-accent-foreground data-[disabled]:pointer-events-none data-[disabled]:opacity-50"
              data-testid="select-item"
              value="banana"
            >
              <span
                class="absolute right-2 flex h-3.5 w-3.5 items-center justify-center"
              >
                <span
                  data-testid="select-item-indicator"
                >
                  <div
                    data-testid="check-icon"
                  />
                </span>
              </span>
              <span
                data-testid="select-item-text"
              >
                Banana
              </span>
            </div>
            <div
              class="-mx-1 my-1 h-px bg-muted"
              data-testid="select-separator"
            />
            <div
              class="relative flex w-full cursor-default select-none items-center rounded-sm py-1.5 pl-2 pr-8 text-sm outline-none focus:bg-secondary focus:text-accent-foreground data-[disabled]:pointer-events-none data-[disabled]:opacity-50"
              data-testid="select-item"
              value="orange"
            >
              <span
                class="absolute right-2 flex h-3.5 w-3.5 items-center justify-center"
              >
                <span
                  data-testid="select-item-indicator"
                >
                  <div
                    data-testid="check-icon"
                  />
                </span>
              </span>
              <span
                data-testid="select-item-text"
              >
                Orange
              </span>
            </div>
          </div>
        </div>
        <div
          class="flex cursor-default items-center justify-center py-1"
          data-testid="select-scroll-down-button"
        >
          <div
            data-testid="chevron-down-icon"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
